File: engine/ios-webview/WebViewEngine/Resource Providers/Providers/BundledResourcesProvider.swift
import Foundation
import UniformTypeIdentifiers

class BundledResourcesProvider: ResourceProvider {
    private let bundle = Bundle(for: BundledResourcesProvider.self)

    func shouldHandle(request: URLRequest) -> Bool {
        request.url?.path().contains("/constellation-mobile-sdk-assets/") ?? false
    }

    func performRequest(_ request: URLRequest) async throws -> (Data, URLResponse) {
        guard let url = request.url else {
            throw LocalProviderError.unexpectedURL
        }

        Log.debug("Local response for \(url)")

        let data = try data(from: extractResourcePath(from: url))
        let response = try createResponse(to: url)
        return (data, response)
    }

    private func extractResourcePath(from url: URL) throws -> String {
        let path = url
            .relativePath
            .range(of: "constellation-mobile-sdk-assets/")
            .map { url.relativePath[$0.upperBound...] }

        if let path, !path.contains("..") {
            return String(path)
        } else {
            throw LocalProviderError.unexpectedURL
        }
    }

    private func data(from path: String) throws -> Data {
        guard let bundleURL = bundle.resourceURL else {
            throw LocalProviderError.fileNotFound
        }

        return try Data(contentsOf: bundleURL.appending(path: path))
    }

    private func createResponse(to requestURL: URL) throws -> URLResponse{
        var headers: [String : String]? = nil

        UTType(filenameExtension: requestURL.pathExtension)?.preferredMIMEType.map {
            headers = ["Content-Type": $0]
        }

        if let response = HTTPURLResponse(
            url: requestURL,
            statusCode: 200,
            httpVersion: nil,
            headerFields: headers
        ) {
            return response
        } else {
            throw LocalProviderError.cannotCreateResponse
        }
    }
}

File: core/src/iosMain/kotlin/com/pega/constellation/sdk/kmp/core/engine/providers/BundledResourcesProvider.kt
package com.pega.constellation.sdk.kmp.core.engine.providers

import com.pega.constellation.sdk.kmp.core.engine.ResourceProvider
import constellation_mobile_sdk.core.generated.resources.Res
import platform.Foundation.*
import platform.UniformTypeIdentifiers.*

class BundledResourcesProvider : ResourceProvider {

    override fun shouldHandle(request: NSURLRequest): Boolean {
        val path = request.URL?.path ?: return false
        return path.contains("/constellation-mobile-sdk-assets/")
    }

    override suspend fun performRequest(request: NSURLRequest): Pair<NSData, NSURLResponse> {
        val url = request.URL ?: throw LocalProviderError.UnexpectedURL
        
        val data = try {
            data(from = "files/${extractResourcePath(url)}")
        } catch (e: Throwable) {
            println("iosMain :: BundledResourceProvider :: Error while performing local request: $e")
            throw LocalProviderError.FileNotFound
        }

        val response = try {
            createResponse(url)
        } catch (e: Throwable) {
            println("iosMain :: BundledResourceProvider :: Error while creating a response for local request: $e")
            throw LocalProviderError.CannotCreateResponse
        }

        return Pair(data, response)
    }

    private fun extractResourcePath(url: NSURL): String {
        val relativePath = url.relativePath ?: throw LocalProviderError.UnexpectedURL
        val marker = "constellation-mobile-sdk-assets/"
        val index = relativePath.indexOf(marker)
        if (index >= 0) {
            val path = relativePath.substring(index + marker.length)
            if (!path.contains("..")) {
                return path
            }
        }
        throw LocalProviderError.UnexpectedURL
    }

    private fun data(from: String): NSData {
        val resUri = Res.getUri(from)
        return NSData.dataWithContentsOfURL(NSURL(string = resUri)) ?: throw LocalProviderError.FileNotFound
    }

    private fun createResponse(requestURL: NSURL): NSURLResponse {
        val headers = mutableMapOf<String, String>()
        requestURL.pathExtension?.let { ext ->
            UTType.typeWithFilenameExtension(ext)?.preferredMIMEType?.let { mimeType ->
                headers.put("Content-Type", mimeType)
            }
        }

        try {
            return NSHTTPURLResponse(
                requestURL,
                200,
                null,
                headers.toMap()
            )
        } catch (_: Throwable) {
            // Just in case failable constructor fails (Kotlin does not see it's failable)
            throw LocalProviderError.CannotCreateResponse
        }

    }
}

sealed class LocalProviderError : Throwable() {
    object UnexpectedURL : LocalProviderError()
    object FileNotFound : LocalProviderError()
    object CannotCreateResponse : LocalProviderError()
}
